### DISCUSSION TITLE
Extend `.model_validate` function, accepting some fields not in object

### DISCUSSION THREAD

**Discussion by mmzeynalli** [pydantic/pydantic#9339] [Question]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Why am I asking this:

I have a SQLAlchemy 2.0 query, which has User fields selected, as well as two aggregated fields, which are labeled. As, it is not possible to access those labelled attributes when used `scalars()`, I am forced to retrieve result as `Row` object. The query also contains `joinedload` of other models, and they have corresponding field in pydantic schema. However, my two labelled aggregate data comes as separate, detached from `User` model:

```Python
async def get_full_user_profile_query(user_id: int):
    return (
        select(
            User,
            cast(func.array_agg(func.distinct(Skill.name)), String).label('skills'),
            cast(func.array_agg(func.distinct(Tool.name)), String).label('tools'),
        )
        .where(User.id == user_id)
        .outerjoin(UserSkill, User.id == UserSkill.user_id)
        .outerjoin(Skill, UserSkill.skill_id == Skill.id)
        .outerjoin(UserTool, User.id == UserTool.user_id)
        .outerjoin(Tool, UserTool.tool_id == Tool.id)
        .options(joinedload(User.profile))
        .group_by(User)
    )

class UserProfileSchema(BaseModel):
    id: int
    first_name: Optional[str]
    last_name: Optional[str]
    avatar: Optional[str]
    email: EmailStr
    profile: 'UserProfileMinimalSchema'
    skills: str
    tools: str

    
async def get_user(user_id: int, db: DbDependency):
    """get user route"""
    query = await get_full_user_profile_query(user_id)
    res = db.execute(query).unique()

    item: Row = unwrap_scalars(res)[0]. # result is: Row(User, skills, tools)
    extra = item._asdict()  # main_obj + labelled aggs
    obj = extra.pop(item._parent._keys[0])
    
    ###### PROBLEM HERE ######
    return UserProfileSchema.model_validate(obj, **extra, from_attributes=True)
```
    
As seen, skills and tools are not part of `User` object in `Row` result, however, I want to pass them directly to `model_validate` as a part of `User` model. I can use `for` loop and `setattr`, however, I find that inefficient, especially If I have many aggregated/labelled fields in `select`.

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [X] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
@mmzeynalli,

Thanks for your question. Could you get your `obj` and `extra` data in dictionary format, then just unpack both of those? Moving this to a discussion, as it falls under the question category!

**Reply by mmzeynalli** [reply]
I tried that, however, considering the user object has its complex structures (a.k.a. profile), it gives an error, as `user.profile` is a `UserProfile` object, not, `UserProfileMinimalSchema` or dict. So, that is not a solution. However, I was able to find a workaround, not a good one though (I mentioned this in a previous one):

```python
# The beginning of the code
obj.__dict__.update(extra)
return UserProfileSchema.model_validate(obj, from_attributes=True)
```

**Comment by mzealey** [comment]
Agree that it would be nice to have an `extra={dict}` option on model_validate (perhaps only when from_attributes is specified) to allow adding extra arbitrary details in that are not in the ORM model.

My workaround at the moment:

```
obj = ... from orm ...
obj.extra_field = 'abc'  # type: ignore[attr-defined] # Hacky way to get it into pydantic
Model.model_validate(obj, from_attributes=True)
```

Which IMO is not very nice.
